In [1]:
import pandas as pd
import numpy as np

In [9]:
orgData = pd.read_csv('/content/drive/MyDrive/Mini Project/Test_Signal_Data.csv')
orgData = pd.DataFrame(orgData.iloc[:,1:5])
orgData['Target Class'] = 0
orgData

Fc         Tc         Tr         Tj  Target Class
0      0.500  32.111391  35.444293  31.247891             0
1      0.500  32.263773  34.946917  31.552655             0
2      0.500  32.136788  35.413207  31.603449             0
3      0.500  32.416155  35.351035  31.933611             0
4      0.500  32.111391  35.413207  31.578052             0
...      ...        ...        ...        ...           ...
10295  0.125  31.882817  59.473766  40.543212             0
10296  0.125  31.527258  59.411594  40.847977             0
10297  0.125  31.552655  59.411594  40.746389             0
10298  0.125  31.425670  59.038562  40.720992             0
10299  0.125  31.400273  59.100734  40.924168             0

[10300 rows x 5 columns]

In [7]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [11]:
dupData = orgData
# def syntheticData(dupData):
#     Fc = np.random.uniform(orgData['Fc'].min(),orgData['Fc'].max(),30000)
#     Tc = np.random.uniform(orgData['Tc'].min(),orgData['Tc'].max(),30000)
#     Tr = np.random.uniform(orgData['Tr'].min(),orgData['Tr'].max(),30000)
#     Tj = np.random.uniform(orgData['Tj'].min(),orgData['Tj'].max(),30000)

#     for j in range(len(Fc)):
#       newRow = {'Fc':Fc[j],'Tc':Tc[j],'Tr':Tr[j],'Tj':Tj[j],'Target Class':0}
#       dupData = dupData.append(newRow,ignore_index=True)
#     return dupData
# dupData = syntheticData(dupData)

In [12]:
dupData

Fc         Tc         Tr         Tj  Target Class
0      0.500  32.111391  35.444293  31.247891             0
1      0.500  32.263773  34.946917  31.552655             0
2      0.500  32.136788  35.413207  31.603449             0
3      0.500  32.416155  35.351035  31.933611             0
4      0.500  32.111391  35.413207  31.578052             0
...      ...        ...        ...        ...           ...
10295  0.125  31.882817  59.473766  40.543212             0
10296  0.125  31.527258  59.411594  40.847977             0
10297  0.125  31.552655  59.411594  40.746389             0
10298  0.125  31.425670  59.038562  40.720992             0
10299  0.125  31.400273  59.100734  40.924168             0

[10300 rows x 5 columns]

In [13]:
def introducing_Error(dupData):
  columns = ['Fc','Tc','Tr','Tj']

  for cols in columns:
    errors = []
    err_number = 1
    if cols == 'Fc':
      for i in range(1000):
        errors.append(0)
      while err_number <= 9000:
        err = np.random.uniform(0,orgData[cols].max()+0.4)
        if float(err) < orgData[cols].min() or float(err) > orgData[cols].max():
          if err not in errors:
            errors.append(err)
            err_number += 1
          else:
            continue
      arr1 = np.random.uniform(orgData['Tc'].min(),orgData['Tc'].max(),10000)
      arr2 = np.random.uniform(orgData['Tr'].min(),orgData['Tr'].max(),10000)
      arr3 = np.random.uniform(orgData['Tj'].min(),orgData['Tj'].max(),10000)
      if len(errors) != 0 :
        for j in range(len(errors)):
          newRow = {'Fc':errors[j],'Tc':arr1[j],'Tr':arr2[j],'Tj':arr3[j],'Target Class':1}
          dupData = dupData.append(newRow,ignore_index=True)
    else:
      while err_number <= 10000:
        err = np.random.uniform(orgData[cols].min()-5,orgData[cols].max()+5)
        if float(err) < orgData[cols].min() or float(err) > orgData[cols].max() and float(err) != 0:
          if err not in errors:
              errors.append(err)
              err_number += 1
          else:
            continue

      if cols == 'Tc':
        arr1 = np.random.uniform(orgData['Fc'].min(),orgData['Fc'].max(),10000)
        arr2 = np.random.uniform(orgData['Tr'].min(),orgData['Tr'].max(),10000)
        arr3 = np.random.uniform(orgData['Tj'].min(),orgData['Tj'].max(),10000)
        if len(errors) != 0 :
          for j in range(len(errors)):
            newRow = {'Fc':arr1[j],'Tc':errors[j],'Tr':arr2[j],'Tj':arr3[j],'Target Class':2}
            dupData = dupData.append(newRow,ignore_index=True)

      elif cols == 'Tr':
        arr1 = np.random.uniform(orgData['Tc'].min(),orgData['Tc'].max(),10000)
        arr2 = np.random.uniform(orgData['Fc'].min(),orgData['Fc'].max(),10000)
        arr3 = np.random.uniform(orgData['Tj'].min(),orgData['Tj'].max(),10000)
        if len(errors) != 0:
          for j in range(len(errors)):
            newRow = {'Fc':arr2[j],'Tc':arr1[j],'Tr':errors[j],'Tj':arr3[j],'Target Class':3}
            dupData = dupData.append(newRow,ignore_index=True)

      elif cols == 'Tj':
        arr1 = np.random.uniform(orgData['Tc'].min(),orgData['Tc'].max(),10000)
        arr2 = np.random.uniform(orgData['Tr'].min(),orgData['Tr'].max(),10000)
        arr3 = np.random.uniform(orgData['Fc'].min(),orgData['Fc'].max(),10000)
        if len(errors) != 0 :
          for j in range(len(errors)):
            newRow = {'Fc':arr3[j],'Tc':arr1[j],'Tr':arr2[j],'Tj':errors[j],'Target Class':4}
            dupData = dupData.append(newRow,ignore_index=True)

  return dupData

In [ ]:
errorData = introducing_Error(dupData)

In [ ]:
errorData

In [ ]:
errorData[errorData['Fc']== 0]

Fc         Tc         Tr         Tj  Target Class
40300  0.0  33.390724  60.686825  39.112949           1.0
40301  0.0  31.054626  74.482677  33.480723           1.0
40302  0.0  31.097670  46.740718  45.494865           1.0
40303  0.0  32.296907  86.132084  55.194174           1.0
40304  0.0  32.328155  38.961177  49.302228           1.0
...    ...        ...        ...        ...           ...
41295  0.0  32.638714  56.809373  34.931804           1.0
41296  0.0  31.892163  72.207935  35.308879           1.0
41297  0.0  31.965048  89.166512  34.511942           1.0
41298  0.0  31.027798  59.433521  52.689782           1.0
41299  0.0  32.014958  69.183316  34.193574           1.0

[1000 rows x 5 columns]

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
errorData[['Fc','Tc','Tr','Tj']] = sc.fit_transform(errorData[['Fc','Tc','Tr','Tj']])
errorData

Fc        Tc        Tr        Tj  Target Class
0     -0.250503 -0.071811 -1.515086 -1.408677           0.0
1     -0.250503 -0.010819 -1.542617 -1.377828           0.0
2     -0.250503 -0.061645 -1.516806 -1.372687           0.0
3     -0.250503  0.050174 -1.520248 -1.339267           0.0
4     -0.250503 -0.071811 -1.516806 -1.375257           0.0
...         ...       ...       ...       ...           ...
80295 -1.014081 -0.615060  1.247828  1.796433           4.0
80296 -1.282919 -1.004882 -1.316288 -1.553658           4.0
80297 -0.876737 -0.668131  1.385917  1.558947           4.0
80298 -0.805372 -0.319418 -0.286029  1.859585           4.0
80299 -0.631837  0.390413 -0.136352  1.774919           4.0

[80300 rows x 5 columns]

In [ ]:
# !pip install imblearn

In [ ]:
X = errorData.iloc[:,[0,1,2,3]]
y = errorData.iloc[:,4]

from imblearn.under_sampling import RandomUnderSampler
rus = RandomUnderSampler(sampling_strategy='majority')
X,y = rus.fit_resample(X,y)
print("rev index:",X.shape)

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)

rev index: (50000, 4)


In [ ]:
from sklearn.svm import SVC
svm = SVC()
param_grid = {'C': [0.1, 1, 10, 100, 1000],
              'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
              'kernel': ['rbf','linear','poly']}

from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(svm,param_grid)
grid.fit(X_train,y_train)
grid_svc = grid.predict(X_test)

met.accuracy_score(y_test,grid_svc)
print(grid.best_estimator_)

In [ ]:
svm.get_params()

{'C': 100,
 'break_ties': False,
 'cache_size': 200,
 'class_weight': None,
 'coef0': 0.0,
 'decision_function_shape': 'ovr',
 'degree': 3,
 'gamma': 0.001,
 'kernel': 'rbf',
 'max_iter': -1,
 'probability': False,
 'random_state': None,
 'shrinking': True,
 'tol': 0.001,
 'verbose': False}

In [ ]:
from sklearn.svm import SVC
classifier = SVC(kernel='rbf')
classifier.fit(X_train,y_train)
y_pred = classifier.predict(X_test)
from sklearn import metrics as met
acc = met.accuracy_score(y_test,y_pred)
acc*100


97.815

In [ ]:
cm = met.confusion_matrix(y_test,y_pred)
cm

array([[2752,   42,    1,  127,   95],
       [  21, 2990,    0,    9,    0],
       [  28,    7, 2942,    6,    2],
       [   0,    0,    0, 2995,    0],
       [   0,    0,    0,    3, 2980]])

 Fc    0.998277,
 Tc    35.286022,
 Tr    90.839534,
 Tj    60.022752

 Fc     0.125000,
 Tc    29.317715,
 Tr    34.946917,
 Tj    30.485979


In [ ]:
input = [[0.5,33,40,96]]
print(svm.predict(input))

[2.]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


In [ ]:
errorData.iloc[:,:4].to_csv('/content/drive/MyDrive/Mini Project/Book1.xlsx',index=False)

In [ ]:
import time
csv_file = orgData
chunk_size = 20
csv_reader = pd.read_csv(csv_file, chunksize=chunk_size)

for chunk in csv_reader:
    print(svm.predict(chunk))
    time.sleep(1)